File: 1315-sum-of-nodes-with-even-valued-grandparent/1315-sum-of-nodes-with-even-valued-grandparent.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def sumEvenGrandparent(self, root: TreeNode) -> int:
        self.tot = 0
        
        def add(root):
            total = 0
            if not root:
                return 0
            if root.left:
                total += root.left.val
            if root.right:
                total += root.right.val
            return total
        
        def dfs(root):
            if not root:
                return
            if root.val % 2 == 0:
                self.tot += add(root.left) + add(root.right)
            dfs(root.left)
            dfs(root.right)
        
        dfs(root)
        return self.tot

File: 1315-sum-of-nodes-with-even-valued-grandparent/1315-sum-of-nodes-with-even-valued-grandparent.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int total = 0;
    int dfs(TreeNode* root, TreeNode* par, TreeNode* gpar) {
        if (!root)
            return 0;
        if (gpar && gpar->val % 2 == 0)
            total += root->val;
        dfs(root->left, root, par);
        dfs(root->right, root, par);
        return total;
    }
    int sumEvenGrandparent(TreeNode* root) {
       return dfs(root, NULL, NULL);
    }
};